### PR TITLE
[SW-1202] Clean PySparkling

### PR DESCRIPTION
**Pull request by jakubhava**
-> Don't need to reference SPARK_HOME to build PySparkling & run unit tests
-> Automatically create virtualenv during build
-> Removes unused code
-> Speeds up the tests
-> Remove hard-coded references to specific Spark. Generate those during build. Simplifies upgrades to a new Spark versions
-> Use proper python plugin to invoke scripts

### CODE REVIEW

File: py/build.gradle [hunk 1]
```diff
@@ -417,12 +431,8 @@ task buildPython(dependsOn: distPython)
...
-                    defaultEnv
-    ].find { h -> h!=null } // first match
-    // Return env
-    logger.info("* Detected '$denv' Sparkling test environment (configure via property 'sparklingTestEnv')")
-    denv
-}
+
+
+
+
```
Comment (mn-mikke) [inline py/build.gradle]: nit: extra lines

File: py/build.gradle [hunk 2]
```diff
@@ -10,6 +12,29 @@ ext {
...
 }
 
+def getPythonVersion() {
+    Python p = new Python(project, python.getPythonPath(), python.getPythonBinary())
+    return p.version
+}
+
+def getPythonMajorVersion() {
+    Python p = new Python(project, python.getPythonPath(), python.getPythonBinary())
+    String[] split = p.version.split("\\.")
```
Comment (mn-mikke) [inline py/build.gradle]: Would it be possible to use ```tokenize(".")``` instead of ```split("\\.")```?

**Comment by jakubhava** [comment]
WIP

File: py/.gitignore [hunk 3]
```diff
@@ -7,10 +7,7 @@ examples/notebooks/.ipynb_checkpoints
...
-# Distribution / packaging
+# Build directory
 build/
-dist/
-*.egg-info/
-version.txt
-sparkling_water
-h2o
\ No newline at end of file
+
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply py/.gitignore]: 👍

File: py/build.gradle (hunk 1 above)
Comment (jakubhava) [reply py/build.gradle]: 👍

File: py/build.gradle (hunk 2 above)
Comment (jakubhava) [reply py/build.gradle]: yup!

File: gradle/python.gradle [hunk 4]
```diff
@@ -1,29 +1,9 @@
...
-  testRuntime {
-    extendsFrom test
-  }
-}
-
-task compile(group: "Build", description: "Compile this project")
-
 task test(group: "Verification", description: "Test this project")
 
 task uploadArchives(group: "Release", description: "Upload all archives")
```
Comment (jakubhava) [inline gradle/python.gradle]: Remove - not used

File: gradle/utils.gradle [hunk 5]
```diff
@@ -36,11 +36,23 @@ def h2oJarLocation() {
...
+def detectTestEnvironment(defaultEnv = "local") {
+    def denv = [ project.hasProperty("sparklingTestEnv") ? project["sparklingTestEnv"] : null,
+                 System.properties["sparklingTestEnv"],
+                 defaultEnv
+    ].find { h -> h!=null } // first match
+    // Return env
+    logger.info("* Detected '$denv' Sparkling test environment (configure via property 'sparklingTestEnv')")
+    denv
+}
+
```
Comment (jakubhava) [inline gradle/utils.gradle]: extra line

File: examples/build.gradle [hunk 6]
```diff
@@ -99,4 +99,7 @@ scriptsTest {
 
   // Show standard out and standard error of the test JVM(s) on the console
   testLogging.showStandardStreams = true
-}
\ No newline at end of file
+}
+
+integTest.dependsOn checkSparkVersion
+scriptsTest.dependsOn checkSparkVersion
```
Comment (jakubhava) [inline examples/build.gradle]: new line at the end of file

File: gradle/spark.gradle [hunk 7]
```diff
@@ -131,11 +131,4 @@ task checkSparkVersionTask {
...
     }
-}
-
-afterEvaluate { project ->
-    Task task = project.tasks.findByName("check")
-    if (task != null) {
-        task.dependsOn checkSparkVersionTask
-    }
-}
+}
```
Comment (jakubhava) [inline gradle/spark.gradle]: nl

File: jenkins/Dockerfile [hunk 8]
```diff
@@ -148,6 +148,18 @@ RUN cd /home/jenkins && \
     ./gradlew build testClasses resolveTestRuntimeDependencies -PtestMojoPipeline=true -x check -x :sparkling-water-r:build -x :sparkling-water-py:build && \
     cd ..
 
+### Create VirtualEnvs for each Spark & Python combination
+RUN cd /home/jenkins/sparkling-water && \
```
Comment (jakubhava) [inline jenkins/Dockerfile]: Loop on Dockerfile level ?

File: py/.gitignore [hunk 9]
```diff
@@ -7,10 +7,5 @@ examples/notebooks/.ipynb_checkpoints
...
-# Distribution / packaging
-build/
-dist/
-*.egg-info/
-version.txt
-sparkling_water
-h2o
\ No newline at end of file
+# Build directory
+build/
```
Comment (jakubhava) [inline py/.gitignore]: nl

File: py/build.gradle [hunk 10]
```diff
@@ -245,140 +251,123 @@ else:
...
 
 artifacts {
     sdist distPython
 }
 
-task testPythonConf(type: Exec, dependsOn: distPython){
-    doFirst {
-        preparePythonEnv(environment)
+def createTestArgs() {
+    def args = []
```
Comment (jakubhava) [inline py/build.gradle]: init

File: py/build.gradle [hunk 11]
```diff
@@ -245,140 +251,123 @@ else:
...
+// Run python integration tests. On Purpose, don't use PythonTask as we want to test the real user
+// use case of submitting apps via ./spark-submit
 //
-task integTestPython(type: Exec, dependsOn: distPython) {
-    doFirst {
-        preparePythonEnv(environment)
-    }
-    def testEnv = detectEnvironment()
+task integTestPython(type: Exec, dependsOn: [distPython]) {
+    def testEnv = detectTestEnvironment()
```
Comment (jakubhava) [inline py/build.gradle]: prepare env

File: jenkins/Dockerfile (hunk 8 above)
Comment (jakubhava) [reply jenkins/Dockerfile]: We can do loop on shell, but I think that will hurt readability

File: gradle/utils.gradle [hunk 12]
```diff
@@ -33,14 +33,25 @@ def detectExternalBackendStartMode(defaultMode = "manual") {
...
 // Export methods by turning them into closures
-ext{
+ext {
     getOsSpecificCommandLine = this.&getOsSpecificCommandLine
     isWindowsBased = this.&isWindowsBased
     detectBackendClusterMode = this.&detectBackendClusterMode
     detectExternalBackendStartMode = this.&detectExternalBackendStartMode
     h2oJarLocation = this.&h2oJarLocation
+    detectTestEnvironment = this.&detectTestEnvironment
 }
```
Comment (mn-mikke) [inline gradle/utils.gradle]: new line?

Comment (jakubhava) [reply gradle/utils.gradle]: done

**Pull request merged**